compiler-rt/lib/radsan/radsan_preinit.cpp

```diff
@@ -16,6 +16,6 @@
 // This code is linked into the main executable when -fsanitize=realtime is in
 // the link flags. It can only use exported interface functions.
```

Comment:
Does this definitely still work on the Linux image? The comment here suggests that the function `radsan::radsan_init` (which no longer has the `RADSAN_EXPORT/SANITIZER_INTERFACE_ATTRIBUTE`) can't be used - is the comment a mistake, maybe?

Reply:
Great question. I have not run this on CI, so I'm not sure!


Perfect time to ask this question (and I'll do my own research too).


1. What is the significance of `extern C`? What functions should be in a block that has that? 
2. What is the significance of the `SANITIZER_INTERFACE_ATTRIBUTE` and what functions should have it?


I was blindly pattern matching in `radsan.h`, and don't have a good mental model of these two things. It would be good for this review to be principled in those decisions as opposed to pattern matching. Wondering if you have any thoughts!

Reply:
Alright, to put in some of my own discovered answers here (please add on or correct if I'm wrong). I'm typing this out mostly for my own education


1. Extern C

Free functions in C++ cannot be automatically compiled and used in C++. The main reason for this is that C does NOT support function overloading.

To support function overloading in C++, the compiler adds a bunch of extra information to the name of the function to differentiate them from other overloads. This process is called "name mangling".

For instance, we the programmer may see:

```
void foo(int x, float y);
void foo(bool b);
```

The compiler/linker will see something else, for example (completely made up)
```
__fooIntFlo209ca
__fooBool29ri04
```

So, what do we do when we need to use a free function that is in our C++ code base in C? we need to make sure the name does not get mangled. This is where `extern "C"` comes in.

`extern "C"` ensures the code inside of it is compiled with the C linkage (instead of the C++ linkage) and ensures no mangling takes place.  These functions can be called from either C or C++ with no issues. 


The functions we should put in an `extern C` block are the functions that need to be called from C world. 

FIXME IN THIS REVIEW: For example I think our init function MUST BE extern C, as it needs to be loaded into the preinit array. 

This comment:

> Does this definitely still work on the Linux image? The comment here suggests that the function radsan::radsan_init (which no longer has the RADSAN_EXPORT/SANITIZER_INTERFACE_ATTRIBUTE) can't be used - is the comment a mistake, maybe?

Seems to make sense now, and I would be surprised if this code ran as written on our linux box.

Reply:
Fixed this in the latest review, all methods called from linux land are linked in the C linkage instead of C++.